mono/metadata/w32file-unix.c

```diff
@@ -2979,146 +2662,346 @@ mono_w32file_get_std_handle (gint stdhandle)
 		g_assert_not_reached ();
 	}
 
-	handle = GINT_TO_POINTER (fd);
+	mono_coop_mutex_lock (&fds_mutex);
+
+	if (g_hash_table_lookup (fds, GINT_TO_POINTER(fd))) {
+		mono_coop_mutex_unlock (&fds_mutex);
+	} else {
+		gpointer handle;
 
-	mono_coop_mutex_lock (&stdhandle_mutex);
+		mono_coop_mutex_unlock (&fds_mutex);
 
-	ok = mono_w32handle_lookup (handle, MONO_W32HANDLE_CONSOLE,
-				  (gpointer *)&file_handle);
-	if (ok == FALSE) {
-		/* Need to create this console handle */
 		handle = _wapi_stdhandle_create (fd, name);
-		
 		if (handle == INVALID_HANDLE_VALUE) {
 			mono_w32error_set_last (ERROR_NO_MORE_FILES);
-			goto done;
+			return INVALID_HANDLE_VALUE;
 		}
 	}
-	
-  done:
-	mono_coop_mutex_unlock (&stdhandle_mutex);
 
-	return(handle);
+	return GINT_TO_POINTER (fd);
 }
 
 gboolean
 mono_w32file_read (gpointer handle, gpointer buffer, guint32 numbytes, guint32 *bytesread)
 {
-	switch (mono_w32handle_get_type (handle)) {
-	case MONO_W32HANDLE_FILE:
-		return file_read(handle, buffer, numbytes, bytesread);
-	case MONO_W32HANDLE_CONSOLE:
-		return console_read(handle, buffer, numbytes, bytesread);
-	case MONO_W32HANDLE_PIPE:
-		return pipe_read(handle, buffer, numbytes, bytesread);
+	FDData *fd_data;
+	gboolean ret;
+
+	mono_coop_mutex_lock (&fds_mutex);
+
+	if (!g_hash_table_lookup_extended (fds, handle, NULL, (gpointer*) &fd_data)) {
+		mono_coop_mutex_unlock (&fds_mutex);
+
+		g_warning ("%s: error looking up fd %p", __func__, GPOINTER_TO_INT(handle));
+		mono_w32error_set_last (ERROR_INVALID_HANDLE);
+		return FALSE;
+	}
+
+	mono_refcount_inc (fd_data);
+
+	mono_coop_mutex_unlock (&fds_mutex);
+
+	switch (fd_data->type) {
+	case FDTYPE_FILE:
+		ret = file_read(fd_data, buffer, numbytes, bytesread);
+		break;
+	case FDTYPE_CONSOLE:
+		ret = console_read(fd_data, buffer, numbytes, bytesread);
+		break;
+	case FDTYPE_PIPE:
+		ret = pipe_read(fd_data, buffer, numbytes, bytesread);
+		break;
 	default:
 		mono_w32error_set_last (ERROR_INVALID_HANDLE);
+		mono_refcount_dec (fd_data);
 		return FALSE;
 	}
+
+	mono_refcount_dec (fd_data);
+	return ret;
 }
 
 gboolean
 mono_w32file_write (gpointer handle, gconstpointer buffer, guint32 numbytes, guint32 *byteswritten)
 {
-	switch (mono_w32handle_get_type (handle)) {
-	case MONO_W32HANDLE_FILE:
-		return file_write(handle, buffer, numbytes, byteswritten);
-	case MONO_W32HANDLE_CONSOLE:
-		return console_write(handle, buffer, numbytes, byteswritten);
-	case MONO_W32HANDLE_PIPE:
-		return pipe_write(handle, buffer, numbytes, byteswritten);
+	FDData *fd_data;
+	gboolean ret;
+
+	mono_coop_mutex_lock (&fds_mutex);
+
+	if (!g_hash_table_lookup_extended (fds, handle, NULL, (gpointer*) &fd_data)) {
+		mono_coop_mutex_unlock (&fds_mutex);
+
+		g_warning ("%s: error looking up fd %p", __func__, GPOINTER_TO_INT(handle));
```

Comment:
IIRC it's possible to create a SafeFileHandle and a filestream from a bogus handle. See https://msdn.microsoft.com/en-us/library/ms143395(v=vs.110).aspx for example